docs/tutorials/index.md

```diff
@@ -0,0 +1,186 @@
+    re-partitions by hashing the key), *broadcast()*, or *rebalance()* (which
+    re-partitions randomly). In a *redistributing* exchange the ordering among
+    the elements is only preserved within each pair of sending and receiving
+    subtasks (for example, subtask[1] of *map()* and subtask[2] of
+    *keyBy/window*). So, for example, the redistribution between the keyBy/window and
+    the Sink operators shown above introduces non-determinism regarding the 
+    order in which the aggregated results for different keys arrive at the Sink.
+
+{% top %}
+
+## Timely Stream Processing
+
+For most streaming applications it is very valuable to be able re-process historic data with the
+same code that is used to process live data -- and to produce deterministic, consistent results,
+regardless.
+
+It can also be crucial to pay attention to the order in which events occurred, rather than the order
+in which they are delivered for processing, and to be able to reason about when a set of events is
+(or should be) complete. For example, consider the set of events involved in an e-commerce
+transaction, or financial trade.
+
+These requirements for timely stream processing can be met by using event time timestamps that are
+recorded in the data stream, rather than using the clocks of the machines processing the data.
+
+{% top %}
+
+## Stateful Stream Processing
+
+Flink's operations can be stateful. This means that how one event is handled can depend on the
+accumulated effect of all the events that came before it. State may be used for something simple,
+such as counting events per minute to display on a dashboard, or for something more complex, such as
+computing features for a fraud detection model.
+
+A Flink application is run in parallel on a distributed cluster. The various parallel instances of a
+given operator will execute independently, in separate threads, and in general will be running on
+different machines.
+
+The set of parallel instances of a stateful operator is effectively a sharded key-value store. Each
+parallel instance is responsible for handling events for a specific group of keys, and the state for
+those keys is kept locally.
+
+The diagram below shows a job running with a parallelism of two across the first three operators in
+the job graph, terminating in a sink that has a parallelism of one. The third operator is stateful,
+and you can see that a fully connected network shuffle is occurring between the second and third
+operators. This is being done to partition the stream by some key, so that all of the events that
+need to be processed together, will be.
+
+<img src="{{ site.baseurl }}/fig/parallel-job.png" alt="State is sharded" class="offset" width="65%" />
+
+State is always accessed locally, which helps Flink applications achieve high throughput and
+low-latency. You can choose to keep state on the JVM heap, or if it is too large, in efficiently
+organized on-disk data structures. 
```

Comment:
```suggestion
low-latency. You can choose to keep state on the JVM heap, or if it is too large, in
efficiently-organized on-disk data structures. 
```

Reply:
No, I think in this case the hyphen goes too far.

Reply:
Double-checked with my wife, the editor. She says no hyphen in this case.

Reply:
ok, thanks - I'm always confused with that in English...even though the "efficiently" relates to "organized" and not "data structures"?
Examples how I would derive that otherwise:
- [efficient] [on-disk] data structures
- efficiently-organized [on-disk] data structures